Add String.indexOf() minibench

dhong44 opened the pull request:
Add minibench for String.indexOf()

Signed-off-by: Daniel Hong <daniel.hong@live.com>

dhong44 committed:
Add String.indexOf() minibench

Add minibench for String.indexOf()

Signed-off-by: Daniel Hong <daniel.hong@live.com>

fjeremic commented on net/adoptopenjdk/bumblebench/string/StringIndexOf.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * 
+ * on z13 and newer z/Architectures with vector facilities enabled, string
+ * indexOf can be accelerated by vector instructions.
```
This is not only a Z statement. It is true for x86 as well as Power. I think we should take the "platform" out of this statement and just say that in general indexOf operation can benefit from vectorization and that this benchmark is meant for testing the performance of such implementations.

dhong44 replied:
7b127ed

dhong44 committed:
Remove platform specific comment

Remove comment in StringIndexOf benchmark that mentions
z architecture, since many architectures can support vectors.

Signed-off-by: Daniel Hong <daniel.hong@live.com>

smlambert commented on net/adoptopenjdk/bumblebench/string/StringIndexOf.java:
```diff
@@ -0,0 +1,62 @@
+package com.ibm.bumblebench.string;
```
net.adoptopenjdk.bumblebench - is the package naming structure now that this is contributed to open project.

Please include the [Apache v2 license header](https://github.com/AdoptOpenJDK/bumblebench/blob/master/net/adoptopenjdk/bumblebench/examples/EmptyBench.java#L1-L13) to any new files.

karianna commented on net/adoptopenjdk/bumblebench/string/StringIndexOf.java:
```diff
@@ -0,0 +1,62 @@
+	private static int STRING_LENGTH = option("stringLength", 0);
+
+	private static String string;
+	private static char[] possibleChars = {'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l',
```
I've seen this block repeated across tests...

dhong44 replied:
This block is used in both my open PR, #4, for `StringConversionBench.java` and `StringIndexOf.java`, here, how would you suggest avoiding this duplicated code?

karianna replied:
Something like a `StringTestData.java`?

karianna commented on net/adoptopenjdk/bumblebench/string/StringIndexOf.java:
```diff
@@ -0,0 +1,62 @@
+
+        for (int j = 0; j < STRING_LENGTH; ++j){
+            sb.append(possibleChars[rand.nextInt(possibleChars.length)]);
+		}
```
whitespace weirdness..

karianna commented on net/adoptopenjdk/bumblebench/string/StringIndexOf.java:
```diff
@@ -0,0 +1,62 @@
+		int length;
+		StringBuilder sb = new StringBuilder(1000);
+
+		rand.setSeed(12345L);
```
We don't care how random this is?

tellison replied:
I expect, to have a repeatable test, the author wanted a predictable sequence of random numbers.

dhong44 replied:
In 851b522 I added an option to make the string pseudorandom. By setting the seed I intended to create a benchmark that could be repeated, but if a user wants the test to be more random they can set the option with `-DBumbleBench.random`

dhong44 committed:
Fix whitespace issues in StringIndexOf

StringIndexOf previously had a mix of both tabs and spaces,
the spaces have been replaced with tabs.

Signed-off-by: Daniel Hong <daniel.hong@live.com>

dhong44 committed:
Add option to make StringIndexOf pseudorandom

Add Pseudorandom option to StringIndexOf benchmark
which will be passed with the option "random".

Signed-off-by: Daniel Hong <daniel.hong@live.com>

dhong44 committed:
Guard against lengths smaller than the index in IndexOf

IndexOf benchmark has options for the length of string
and also the index where we set our search character,
this would have cause an IndexOutOfBoundsException but
the intended behavior would be that the search character
will not be found since it is outside of the String.

Signed-off-by: Daniel Hong <daniel.hong@live.com>

dhong44 committed:
Change name of StringIndexOf to StringIndexOfBench

Change name to include the word Bench in file and class
to follow naming convention of other benchmarks.

Signed-off-by: Daniel Hong <daniel.hong@live.com>

dhong44 committed:
Use common StringTestData

Remove declaration of char array and use common data found
in StringTestData class to avoid duplicated code.

Signed-off-by: Daniel Hong <daniel.hong@live.com>

dhong44 commented:
Prereq #5
